src/app/shared/services/chat/offline/offline-chat.service.ts

```diff
@@ -0,0 +1,99 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Observable, throwError, Subject, of, forkJoin, BehaviorSubject } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { ChatMessage } from '../chat-msg.model';
+import { ChatService } from '../chat.service';
+import { RapidProOfflineFlow } from './chat.flow';
+import { RapidProFlowExport } from './rapid-pro-export.model';
+
+export type FlowStatusChange = {
+  flowId: string;
+  status: "start" | "completed" | "expired";
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class OfflineChatService implements ChatService {
+
+  private inflightRequests: Observable<any>[] = [];
+  private flowsById: { [flowId: string]: RapidProOfflineFlow } = {};
+  private contactFields: { [field: string]: string } = {};
+  private currentFlow: RapidProOfflineFlow;
+
+  public messages$: BehaviorSubject<ChatMessage[]>;
+  private flowStatus$: BehaviorSubject<FlowStatusChange[]>;
+
+  constructor(
+    protected http: HttpClient,
+  ) {
+    this.messages$ = new BehaviorSubject([]);
```

Comment:
personal preference, I often just instantiate in the definition `private messages$ = new BehaviorSubject<ChatMessage[]>([]), but know lots of people who consider that sacrilege and instantiate everything within the constructor. So just to mention in case you weren't aware